### PR TITLE
Parse pools in a Step as dict

### PR DESCRIPTION
**Pull request by tcezard**
Add XmlPooledInputDict that parses the pools from a Step
Also fix a few bug associated with the editing of dictionaries

### CODE REVIEW

File: pyclarity_lims/descriptors.py [hunk 1]
```diff
@@ -292,6 +292,8 @@ def _setitem(self, key, value):
         self._elems[0].attrib[key] = value
 
     def _delitem(self, key):
+        if key in ['artifact', 'step', 'rework-step']:
+            key = key + '-uri'
```
Comment (tcezard) [inline pyclarity_lims/descriptors.py]: Deleting any of `['artifact', 'step', 'rework-step']` 's key would  have resulted in `KeyError`

File: pyclarity_lims/descriptors.py [hunk 2]
```diff
@@ -312,13 +314,20 @@ def _parse_element(self, element, **kwargs):
     def _setitem(self, key, value):
         if not isinstance(key, str):
             raise ValueError()
+        elem1 = None
+        for node in self._elems:
+            if node.find('value').text == key:
+                elem1 = node
+                break
+        if elem1:
+            self.rootnode(self.instance).remove(elem1)
```
Comment (tcezard) [inline pyclarity_lims/descriptors.py]: Creating a placement was working however overwriting was resulting in a new placement being created and the old one left intact

File: pyclarity_lims/descriptors.py [hunk 3]
```diff
@@ -358,6 +370,52 @@ def _delitem(self, key):
...
+        if not isinstance(key, str):
+            raise ValueError()
+        if not isinstance(value, tuple) or not len(value) == 2:
+            raise TypeError('You need to provide a tuple of 2 elements not ' + type(value))
+        pool, list_input = value
+        self._delitem(key)
+        node = ElementTree.SubElement(self.rootnode(self.instance), 'pool')
+        node.attrib['name'] = key
+        node.attrib['uri'] = pool.uri
+        for input in list_input:
```
Comment (katieemelianova) [inline pyclarity_lims/descriptors.py]: Should perhaps not be an already defined Python function (input)

Comment (mwhamgenomics) [inline pyclarity_lims/descriptors.py]: Shadows builtin name `input`

File: pyclarity_lims/descriptors.py [hunk 4]
```diff
@@ -358,6 +370,52 @@ def _delitem(self, key):
...
+            node.append(sub)
+
+    def _delitem(self, key):
+        for node in self._elems:
+            if node.attrib['name'] == key:
+                self.rootnode(self.instance).remove(node)
+                break
+
+    def _parse_element(self, element, **kwargs):
+        from pyclarity_lims.entities import Artifact
```
Comment (mwhamgenomics) [inline pyclarity_lims/descriptors.py]: Does `Artifact` take long enough to import that we need to lazy-load it?

Comment (tcezard) [reply pyclarity_lims/descriptors.py]: This is not about lazy loading but is meant to avoid import cyclic dependencies.

**Pull request merged**